src/retry.rs

```diff
@@ -0,0 +1,450 @@
+/// strategies: [FixedDelay], [FibonacciBackoff], and [ExponentialBackoff].
+///
+/// Example:
+/// ```
+/// use std::io::{Error, ErrorKind};
+/// use std::ops::AddAssign;
+/// use std::time::Duration;
+/// use fluvio_future::retry::FixedDelay;
+/// use fluvio_future::retry::retry;
+///
+/// fluvio_future::task::run(async {
+///     let mut attempts = 0u8;
+///     let result = retry(FixedDelay::from_millis(100).take(2), || {
+///         attempts.add_assign(1);
+///         operation()
+///     }).await;
+///     assert!(matches!(result, Err(err) if err.kind() == ErrorKind::NotFound));
+///     assert_eq!(attempts, 3); // first attempt + 2 retries
+/// });
+///
+/// async fn operation() -> Result<(), Error> {
+///     Err(Error::from(ErrorKind::NotFound))
+/// }
+/// ```
+pub async fn retry_if<I, O, F, E, A, P>(retries: I, mut factory: A, condition: P) -> Result<O, E>
+where
+    I: IntoIterator<Item = Duration>,
+    A: FnMut() -> F,
+    F: Future<Output = Result<O, E>>,
+    P: Fn(&E) -> bool,
+    E: Debug,
+{
+    let mut err = poll_err!(factory, condition);
+    for delay_duration in retries.into_iter() {
+        let _ = sleep(delay_duration).await;
+        warn!(?err, "retrying");
+        err = poll_err!(factory, condition);
+    }
+    Err(err)
+}
+
+mod delay {
+    use std::time::Duration;
+
+    /// A retry strategy driven by a fixed interval between retries.
+    /// ```
+    /// use std::io::Error;
+    /// use fluvio_future::retry::{FixedDelay, retry};
+    /// use std::time::{Duration, Instant};
+    ///
+    /// fluvio_future::task::run(async {
+    ///     let _: Result<(), Error> = retry(FixedDelay::from_millis(100).take(4), || async {Ok(())}).await; // 4 retries
+    /// });
+    /// ```
+    #[derive(Default, Clone, Debug, Eq, PartialEq)]
+    pub struct FixedDelay {
+        delay: Duration,
+    }
+
+    impl FixedDelay {
+        pub fn new(delay: Duration) -> Self {
+            Self { delay }
+        }
+
+        pub fn from_millis(millis: u64) -> Self {
+            Self::new(Duration::from_millis(millis))
+        }
+
+        pub fn from_secs(secs: u64) -> Self {
+            Self::new(Duration::from_secs(secs))
+        }
+    }
+
+    impl Iterator for FixedDelay {
+        type Item = Duration;
+
+        fn next(&mut self) -> Option<Duration> {
+            Some(self.delay)
+        }
+    }
+
+    /// A retry strategy driven by the fibonacci series of intervals between retires.
+    /// ```
+    /// use std::io::Error;
+    /// use fluvio_future::retry::{FibonacciBackoff, retry};
+    /// use std::time::{Duration, Instant};
+    ///
+    /// fluvio_future::task::run(async {
+    ///     let _: Result<(), Error> = retry(FibonacciBackoff::from_millis(100).take(4), || async {Ok(())}).await; // 4 retries
+    /// });
+    /// ```
+    #[derive(Default, Clone, Debug, Eq, PartialEq)]
+    pub struct FibonacciBackoff {
```

Comment:
Instead of creating our own implementation, can we use the existing library?  https://crates.io/crates/adaptive_backoff

Reply:
Our structs not only provide `Duration` for the delay but are also responsible for the number of retries. Plus, we have fixed intervals, Fibonacci and exponential. In that crate, only exponential is available.

Reply:
fair enough